sprint 12/1 p2p review

AlexanderSparrow opened the pull request.

AlexanderSparrow committed:
Финальное задание 11 го спринта.
Расширяем функционал приложения. Доработана модель данных для дальнейшего развития проекта (хранение данных в базе SQL)

добавлена схема БД в README.md

AlexanderSparrow committed:
Финальное задание 11 го спринта.
Расширяем функционал приложения. Доработана модель данных для дальнейшего развития проекта (хранение данных в базе SQL)

добавлена схема БД в README.md

AlexanderSparrow committed:
Финальное задание 11 го спринта.
Расширяем функционал приложения. Доработана модель данных для дальнейшего развития проекта (хранение данных в базе SQL)

добавлена схема БД в README.md

AlexanderSparrow committed:
Промежуточное задание 12 го спринта.
Расширяем функционал приложения. Доработана модель данных для дальнейшего развития проекта (хранение данных в базе SQL)

добавлена схема БД и ее описание в README.md

AlexanderSparrow committed:
Финальное задание 11 го спринта.
Расширяем функционал приложения. Доработана модель данных для дальнейшего развития проекта (хранение данных в базе SQL)

добавлена схема БД в README.md

Irina-Gavrilina commented on README.md:
```diff
@@ -1,2 +1,4 @@
 # java-filmorate
 Template repository for Filmorate project.
+
+![filmarate SQL scheme](https://github.com/AlexanderSparrow/java-filmorate/blob/add-database/src/main/java/ru/yandex/practicum/filmorate/db_scheme/filmarate_SQL_%20scheme.png?raw=true)
```
Лучше добавить небольшое пояснение к схеме согласно ТЗ: можно кратко описать, что представляют из себя таблицы, приложить примеры запросов для основных операций приложения (например, получение всех фильмов/пользователей, топ N наиболее популярных фильмов, общих друзей с другим пользователем и т.д.)

AlexanderSparrow replied:
Да, справедливо. Добавил описание структуры базы, полей таблицы, привел примеры запросов.

Irina-Gavrilina commented on src/main/java/ru/yandex/practicum/filmorate/db_scheme/filmarate_SQL_ scheme.png:
По схеме не понятно есть ли у типа данных VARCHAR в таблицах указание фиксированной длины. Если нет, то лучше этим не пренебрегать, так как указание фиксированной длины позволяет сэкономить ресурсы и обеспечивает хранение данных высокого качества

AlexanderSparrow replied:
Пока в теоретической части было явно указано на ограничение длинных поля description- 200 знаков. Это реализовано. 
В дальнейшем, учту рекомендацию и добавлю ограничений по остальным полям, как будет рекомендоваться в ТЗ. Или с сточки зрения целесообразности.

AlexanderSparrow committed:
Merge remote-tracking branch 'origin/add-database' into add-database

AlexanderSparrow committed:
Промежуточное задание 12 го спринта.
Расширяем функционал приложения. Доработана модель данных для дальнейшего развития проекта (хранение данных в базе SQL)

добавлена схема БД и ее описание в README.md
easy polishing

AlexanderSparrow committed:
Промежуточное задание 12 го спринта.
Расширяем функционал приложения. Доработана модель данных для дальнейшего развития проекта (хранение данных в базе SQL)

добавлена схема БД и ее описание в README.md
easy polishing

Irina-Gavrilina reviewed (approved):
Отличная работа!

AlexanderSparrow committed:
Финальное задание 12 го спринта.
Расширяем функционал приложения. Доработана модель данных для дальнейшего развития проекта (хранение данных в базе SQL)

подключена база H2

AlexanderSparrow committed:
Финальное задание 12 го спринта.
Расширяем функционал приложения. Доработана модель данных для дальнейшего развития проекта (хранение данных в базе SQL)

добавлена schema.sql, data.sql
Создана часть DAO классов

AlexanderSparrow committed:
Финальное задание 12 го спринта.
Расширяем функционал приложения. Доработана модель данных для дальнейшего развития проекта (хранение данных в базе SQL)

добавлена schema.sql, data.sql
Создана часть DAL классов (обработка User логики)

AlexanderSparrow committed:
Финальное задание 12 го спринта.
Расширяем функционал приложения. Доработана модель данных для дальнейшего развития проекта (хранение данных в базе SQL)

добавлена schema.sql, data.sql
Создана часть DAL классов (обработка User логики)

Тесты postman по Гыукг проходят норм,

AlexanderSparrow committed:
Финальное задание 12 го спринта.
Расширяем функционал приложения. Доработана модель данных для дальнейшего развития проекта (хранение данных в базе SQL)

Создана часть DAL классов (обработка MPA логики)

Тесты postman по MPA проходят не все :(

AlexanderSparrow committed:
Финальное задание 12 го спринта.
Расширяем функционал приложения. Доработана модель данных для дальнейшего развития проекта (хранение данных в базе SQL)

Создана часть DAL классов (обработка MPA логики)

Тесты postman по MPA проходят на "Ура":)

AlexanderSparrow committed:
Финальное задание 12 го спринта.
Расширяем функционал приложения. Доработана модель данных для дальнейшего развития проекта (хранение данных в базе SQL)

Создана часть DAL классов (обработка Genre логики)

Тесты postman по Genre проходят на "Ура":)

AlexanderSparrow committed:
Финальное задание 12 го спринта.
Расширяем функционал приложения. Доработана модель данных для дальнейшего развития проекта (хранение данных в базе SQL)

Создана часть DAL классов (обработка Genre логики)

Тесты postman по Genre проходят на "Ура":)

чистка кода для Checkstyle

AlexanderSparrow committed:
Финальное задание 12 го спринта.
Расширяем функционал приложения. Доработана модель данных для дальнейшего развития проекта (хранение данных в базе SQL)

Создана часть DAL классов (обработка Genre логики)

Тесты postman по Genre проходят на "Ура":)

чистка кода для Checkstyle

AlexanderSparrow committed:
Финальное задание 12 го спринта.
Расширяем функционал приложения. Доработана модель данных для дальнейшего развития проекта (хранение данных в базе SQL)

Создана часть DAL классов (обработка Genre логики)

рефакторинг и ловля багов

AlexanderSparrow committed:
Финальное задание 12 го спринта.
Расширяем функционал приложения. Доработана модель данных для дальнейшего развития проекта (хранение данных в базе SQL)

Создана часть DAL классов (обработка Genre логики)

рефакторинг и ловля багов

AlexanderSparrow committed:
Финальное задание 12 го спринта.
Расширяем функционал приложения. Доработана модель данных для дальнейшего развития проекта (хранение данных в базе SQL)

Создана часть DAL классов (обработка Genre логики)

рефакторинг и ловля багов

AlexanderSparrow committed:
Финальное задание 12 го спринта.
Расширяем функционал приложения. Доработана модель данных для дальнейшего развития проекта (хранение данных в базе SQL)

тесты postman прохлдят. Готовим к проверке

AlexanderSparrow committed:
Финальное задание 12 го спринта.
Расширяем функционал приложения. Доработана модель данных для дальнейшего развития проекта (хранение данных в базе SQL)

тесты postman прохлдят. Готовим к проверке

AlexanderSparrow committed:
Финальное задание 12 го спринта.
Расширяем функционал приложения. Доработана модель данных для дальнейшего развития проекта (хранение данных в базе SQL)

тесты postman прохлдят. Готовим к проверке

AlexanderSparrow committed:
Финальное задание 12 го спринта.
Расширяем функционал приложения. Доработана модель данных для дальнейшего развития проекта (хранение данных в базе SQL)

тесты postman прохлдят. Готовим к проверке

AlexanderSparrow committed:
Финальное задание 12 го спринта.
Расширяем функционал приложения. Доработана модель данных для дальнейшего развития проекта (хранение данных в базе SQL)

тесты postman прохлдят. Готовим к проверке

AlexanderSparrow committed:
Финальное задание 12 го спринта.
Расширяем функционал приложения. Доработана модель данных для дальнейшего развития проекта (хранение данных в базе SQL)

тесты postman прохлдят. Готовим к проверке

AlexanderSparrow committed:
Финальное задание 12 го спринта.
Расширяем функционал приложения. Доработана модель данных для дальнейшего развития проекта (хранение данных в базе SQL)

тесты postman прохлдят. Готовим к проверке

AlexanderSparrow committed:
Финальное задание 12 го спринта.
Расширяем функционал приложения. Доработана модель данных для дальнейшего развития проекта (хранение данных в базе SQL)

тесты postman теперь точно проходят:))). Готовим к ревью

AlexanderSparrow committed:
Финальное задание 12 го спринта.
Расширяем функционал приложения. Доработана модель данных для дальнейшего развития проекта (хранение данных в базе SQL)

тесты postman теперь точно проходят:))). Готовим к ревью

AlexanderSparrow committed:
Финальное задание 12 го спринта.
Расширяем функционал приложения. Доработана модель данных для дальнейшего развития проекта (хранение данных в базе SQL)

тесты postman теперь точно проходят, покрыл основные методы хранилища тестами. Готовим к ревью

AlexanderSparrow committed:
Финальное задание 12 го спринта.
Расширяем функционал приложения. Доработана модель данных для дальнейшего развития проекта (хранение данных в базе SQL)

тесты postman теперь точно проходят, покрыл основные методы хранилища тестами. Готовим к ревью (небольшой рефакторин)

AlexanderSparrow committed:
Финальное задание 12 го спринта.
Расширяем функционал приложения. Доработана модель данных для дальнейшего развития проекта (хранение данных в базе SQL)

тесты postman теперь точно проходят, покрыл основные методы хранилища тестами. Готовим к ревью (не все тесты на гите проходят проверку)

AlexanderSparrow committed:
Финальное задание 12 го спринта.
Расширяем функционал приложения. Доработана модель данных для дальнейшего развития проекта (хранение данных в базе SQL)

тесты postman теперь точно проходят, покрыл основные методы хранилища тестами. Готовим к ревью (не все тесты на гите проходят проверку)

AlexanderSparrow committed:
Финальное задание 12 го спринта.
Расширяем функционал приложения. Доработана модель данных для дальнейшего развития проекта (хранение данных в базе SQL)

тесты postman теперь точно проходят, покрыл основные методы хранилища тестами. Готовим к ревью (не все тесты на гите проходят проверку)

AlexanderSparrow committed:
Финальное задание 12 го спринта.
Расширяем функционал приложения. Доработана модель данных для дальнейшего развития проекта (хранение данных в базе SQL)

тесты postman теперь точно проходят, покрыл основные методы хранилища тестами. Готовим к ревью (не все тесты на гите проходят проверку)

AlexanderSparrow committed:
Финальное задание 12 го спринта.
Расширяем функционал приложения. Доработана модель данных для дальнейшего развития проекта (хранение данных в базе SQL)

тесты postman теперь точно проходят, покрыл основные методы хранилища тестами. Готовим к ревью (не все тесты на гите проходят проверку)
Для исключения ошибок по причине нескольких прогонов теста, добавлено создание и инициализация таблиц с liquibase

AlexanderSparrow committed:
Финальное задание 12 го спринта.
Расширяем функционал приложения. Доработана модель данных для дальнейшего развития проекта (хранение данных в базе SQL)

тесты postman теперь точно проходят, покрыл основные методы хранилища тестами. Готовим к ревью (не все тесты на гите проходят проверку)
Удалил таблицу friendship_status, небольшой рефакторинг, поиск багов

AlexanderSparrow committed:
Финальное задание 12 го спринта.
вернул к жизни тест FilmDbStorageTest

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -2,21 +2,31 @@
 
 import java.util.List;
 
+@Log4j2
```
Лучше писать @Slf4j вместо @Log4j2, т.к. это абстракция вместо конкретной библиотеке и тогда вы сможете поменять реализацию в дальнейшем
The Simple Logging Facade for Java (SLF4J) serves as a simple facade or abstraction1 for various logging frameworks (e.g. java.util.logging, logback, log4j) allowing the end user to plug in the desired logging framework at deployment time.

AlexanderSparrow replied:
Спасибо, учту на будущее. Поправил:)

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -2,21 +2,31 @@
+    public ResponseEntity<Film> addFilm(@Valid @RequestBody Film film) {
+        log.info("Получен запрос на создание фильма: {}", film);
+        Film createdFilm = filmService.addFilm(film);
+        if (createdFilm == null || createdFilm.getId() == 0) {
```
Обязательно
Если вы не можете добавить фильм, то нужно кидать ошибку именно в сервисном классе, не нужно добавлять проверки в контроллере

AlexanderSparrow replied:
Удалил проверку

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -29,13 +39,24 @@ public List<Film> getAllFilms() {
-    public void addLike(@PathVariable int filmId, @PathVariable int userId) {
+    public ResponseEntity<Void> addLike(@PathVariable long filmId, @PathVariable long userId) {
         filmService.addLike(filmId, userId);
+        return ResponseEntity.status(HttpStatus.OK).build();
```
Можно лучше
Можно использовать статические методы
ResponseEntity.ok().build()

AlexanderSparrow replied:
Спасибо, учту на будущее. Поправил:)

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/dal/FilmRepository.java:
```diff
@@ -0,0 +1,93 @@
+            PreparedStatement ps = connection.prepareStatement(sql, new String[]{"id"});
+            ps.setString(1, film.getName());
+            ps.setString(2, film.getDescription());
+            ps.setDate(3, java.sql.Date.valueOf(film.getReleaseDate()));
```
Можно лучше
Писать полностью класс с указанием пакета необязательно, можно написать вот так
 ps.setDate(3, Date.valueOf(film.getReleaseDate()));

AlexanderSparrow replied:
Улучшил:)

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/dal/FriendshipRepository.java:
```diff
@@ -0,0 +1,48 @@
+    /*public List<Long> getUserFriends(long userId) {
+        String sql = "SELECT friend_id FROM friendships WHERE user_id = ?";
+        return jdbcTemplate.queryForList(sql, Long.class, userId);
+    }*/
```
Закомментированный код надо удалить, в гите всё останется

AlexanderSparrow replied:
Да, спасибо, не удалил при коммите

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/dal/FriendshipRepository.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    // Получить статус дружбы между пользователями
+    public Integer getFriendshipStatus(long userId, long friendId) {
```
Обязательно
Этот метод нужно удалить, он неверный т.к. нет status_id

AlexanderSparrow replied:
Да, спасибо! После удаления таблица со статусами дружбы можно удалить. Не был уверен, что это причина непрохождения тестов на гит

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/dal/mappers/FilmRowMapper.java:
```diff
@@ -0,0 +1,66 @@
+        film.setDuration(resultSet.getInt("duration"));
+
+        // Маппинг MPA рейтинга
+        Mpa mpa = getMpaRating(resultSet.getInt("mpa_rating_id"));
```
Обычно в мапперах не делаются sql запросы, а подаются необходимые данные для маппинга, т.к. это будет противоречит SOLID, но пусть остается так, но на будущее имейте это в виду

AlexanderSparrow replied:
Перенес запросы в соответсвующие контроллеры

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/model/Friendship.java:
```diff
@@ -0,0 +1,11 @@
+
+    private long userId;      // ID пользователя, который отправил запрос на дружбу
+    private long friendId;    // ID пользователя, которому был отправлен запрос
+    private FriendshipStatus status;  // Статус дружбы (подтверждённая или неподтверждённая)
```
Обязательно
status надо удалить

AlexanderSparrow replied:
Да, спасибо, теперь это лишнее

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/model/FriendshipStatus.java:
```diff
@@ -0,0 +1,9 @@
+import lombok.Data;
+
+@Data
+public class FriendshipStatus {
```
Обязательно
Этот класс надо удалить

AlexanderSparrow replied:
Да, теперь это очевидо

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/model/User.java:
```diff
@@ -25,5 +29,10 @@ public class User {
+
+    // Связи с другими пользователями (друзья) и их статус дружбы
+    //Заготовка для DB
+    private Set<Friendship> friendsAndFriendships = new HashSet<>();
```
Обязательно
У User не должен быть список друзей и в тестах это не проверяется, можно удалить friends и friendsAndFriendships. Нужно будет мапперы подправить

AlexanderSparrow replied:
Убрал, поправил:)

kesch9 replied:
FriendshipRowMapper надо также удалить он не нужен. нигде не используется

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -50,4 +87,27 @@ public List<Film> getPopularFilms(int count) {
+                .orElseThrow(() -> new FilmNotFoundException(filmId));
+    }
+
+    private void validateUniqueGenresAndMpa(Film film) {
```
По моему эта валидация не нужна, попробуйте удалить, если тесты будут падать верните

AlexanderSparrow replied:
Убрал, вроде не падают:)

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -1,47 +1,84 @@
 
     public List<Film> getPopularFilms(int count) {
-        return filmStorage.getAllFilms().stream()
+        return filmStorage.getPopularFilms(count).stream()
```
Обязательно
В запросе у вас уже отсортированно, сортировать в stream не нужно и ограничить можно в самом запросе использовав LIMIT в запросе

AlexanderSparrow replied:
Поправлено

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -1,63 +1,83 @@
 @Service
-@RequiredArgsConstructor
 public class UserService {
+    @Qualifier("userDbStorage")
```
...DbStorage лишние сущности используйте вместо них repostiory

AlexanderSparrow replied:
Рефакторинг выполнен

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -1,63 +1,83 @@
         user.getFriends().add(friendId);  // Добавляем друга пользователю
         log.info("Пользователь {} добавлен в друзья пользователю {}", userId, friendId);
         friend.getFriends().add(userId);  // Добавляем пользователя другу
+        userStorage.addFriend(userId, friendId);;
```
Нужно выполнить только
addFriend
Добавление друга пользователю и обновление пользователя не нужны

AlexanderSparrow replied:
Это было оставлено для сохранения реализации inMemory. Поправил

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -1,63 +1,83 @@
-    public List<User> getCommonFriends(int userId1, int userId2) {
-        log.info("Получение спискаобщих друзей для пользователя {} и пользователя {}", userId1, userId2);
+    public List<User> getCommonFriends(Long userId1, Long userId2) {
+        log.info("Получение списка общих друзей для пользователя {} и пользователя {}", userId1, userId2);
```
Обязательно
Нужно написать запрос получение общих друзей, а не использовать stream
Дам подсказку как должен выглядить запрос
"select * from USERS u, FRIENDS f, FRIENDS o " +
				"where u.USER_ID = f.FRIEND_ID AND u.USER_ID = o.FRIEND_ID AND f.USER_ID = ? AND o.USER_ID = ?";

AlexanderSparrow replied:
Исправил

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -68,11 +88,12 @@ public List<User> getAllUsers() {
     }
 
-    public Set<User> getUserFriends(int userId) {
+    public Set<User> getUserFriends(Long userId) {
```
Обязательно
Нужно написать запрос вида
"select * from USERS, FRIENDS where USERS.USER_ID = FRIENDS.FRIEND_ID AND FRIENDS.USER_ID = ?
а не по id запрашивать каждого друга

AlexanderSparrow replied:
исправил

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmDbStorage.java:
```diff
@@ -0,0 +1,61 @@
+@Repository
+@Qualifier("filmDbStorage")
+@AllArgsConstructor
+public class FilmDbStorage implements FilmStorage {
```
Обязательно
...DbStorage это лишние сущности вместо них используйте FilmRepository

AlexanderSparrow replied:
Рефакторинг выполнен

kesch9 commented on src/main/java/ru/yandex/practicum/filmorate/storage/InMemoryFilmStorage.java:
```diff
@@ -51,14 +49,20 @@ public List<Film> getAllFilms() {
     }
 
+    @Override
+    public void removeLike(long filmId, long likeId) {
```
Обязательно
InMemory.,, реализации нужно удалить и не надо добавлять пустые методы, просто удалите все классы InMemory

AlexanderSparrow replied:
Удалил прежнюю реализацию

kesch9 reviewed (commented):
Нужно еще поработать над кодом

AlexanderSparrow committed:
Финальное задание 12 го спринта.
промежуточный коммит, исправления замечаний ревьюера (без удаления реализации inMemory)

AlexanderSparrow committed:
Финальное задание 12 го спринта.
Рефакторинг кода по рекомендации ревьюера (удалена реализация inMemory), устранены прочие замечания по коду

AlexanderSparrow committed:
Финальное задание 12 го спринта.
Рефакторинг кода по рекомендации ревьюера (удалена реализация inMemory), устранены прочие замечания по коду (финальная подстстка)

AlexanderSparrow committed:
Финальное задание 12 го спринта.
Рефакторинг кода по рекомендации ревьюера (удалена реализация inMemory), устранены прочие замечания по коду (финальная подстстка)

AlexanderSparrow committed:
Финальное задание 12 го спринта.
Рефакторинг кода по рекомендации ревьюера (удалена реализация inMemory), устранены прочие замечания по коду (финальная подчистка и упорядочение тестовых классов)

kesch9 reviewed (commented):
Нужно подправить ошибки checkstyle, чтобы сборка прошла

AlexanderSparrow committed:
Финальное задание 12 го спринта.
Рефакторинг кода по рекомендации ревьюера (удалена реализация inMemory), устранены прочие замечания по коду (финальная подчистка и упорядочение тестовых классов)

AlexanderSparrow commented:
> Нужно подправить ошибки checkstyle, чтобы сборка прошла

досадненько, поправил:)

kesch9 reviewed (approved):
Approve

AlexanderSparrow committed:
доработка application.yml
